usecases/cem/ohpcf/public.go

```diff
@@ -73,10 +181,144 @@ func (e *OHPCF) WriteSmartEnergyManagementData(entity spineapi.EntityRemoteInter
+	}
+
+	startTime, err := sequence.Schedule.StartTime.GetTime()
+	if err != nil {
+		return nil, err
+	}
+
+	return &startTime, nil
+}
+
+// Scenario 2 - Control heat pump compressor's power consumption flexibility
+
+func (e *OHPCF) WriteStartTime(entity spineapi.EntityRemoteInterface, startTime time.Time) error {
+	if !e.IsCompatibleEntityType(entity) {
+		return api.ErrNoCompatibleEntity
+	}
+
+	smartEnergyManagement, err := client.NewSmartEnergyManagementPs(e.LocalEntity, entity)
+	if err != nil || smartEnergyManagement == nil {
+		return api.ErrDataNotAvailable
+	}
+
+	if startTime.Compare(time.Now()) < 0 {
+		logging.Log().Error("Start time is in the past")
+		return api.ErrNotSupported
+	}
+
+	smartEnergyManagementPsData, err := e.getCopyOfSmartEnergyManagementData(entity)
+	if err != nil {
+		return err
+	}
+
+	sequence := &smartEnergyManagementPsData.Alternatives[0].PowerSequence[0]
+	sequence.Schedule = &model.PowerSequenceScheduleDataType{
+		StartTime: model.NewAbsoluteOrRelativeTimeTypeFromDuration(time.Until(startTime)),
+	}
+
+	_, err = smartEnergyManagement.WriteData(smartEnergyManagementPsData)
+
+	return err
+}
+
+func (e *OHPCF) writeState(entity spineapi.EntityRemoteInterface, state model.PowerSequenceStateType) error {
+	if !e.IsCompatibleEntityType(entity) {
+		return api.ErrNoCompatibleEntity
+	}
+
+	smartEnergyManagement, err := client.NewSmartEnergyManagementPs(e.LocalEntity, entity)
+	if err != nil || smartEnergyManagement == nil {
+		return api.ErrDataNotAvailable
+	}
+
+	smartEnergyManagementPsData, err := e.getCopyOfSmartEnergyManagementData(entity)
+	if err != nil {
+		return err
+	}
+
+	sequence := smartEnergyManagementPsData.Alternatives[0].PowerSequence[0]
+	sequence.State.State = &state
+
+	_, err = smartEnergyManagement.WriteData(smartEnergyManagementPsData)
+
+	return err
+}
+
+func (e *OHPCF) Start(entity spineapi.EntityRemoteInterface) error {
+	flexibilitySettings, _ := e.PowerConsumptionFlexibilitySettings(entity)
+	currentState, _ := e.State(entity)
+
+	if flexibilitySettings == nil || currentState == nil {
+		return api.ErrDataNotAvailable
+	}
+
+	switch *currentState {
+	case ucapi.PowerSequenceStateTypeInactive:
+		// Schedule (activate) optional power consumption process to begin within 5 seconds
+		return e.WriteStartTime(entity, time.Now().Add(time.Second*5))
+
+	case ucapi.PowerSequenceStateTypeRunning:
+		// Already running
+		return nil
+
+	case ucapi.PowerSequenceStateTypePaused:
+		if flexibilitySettings.IsPauseable {
+			// Resume the process
+			return e.writeState(entity, model.PowerSequenceStateTypeRunning)
+		}
+	}
+
+	return api.ErrNotSupported
+}
+
+func (e *OHPCF) Stop(entity spineapi.EntityRemoteInterface) error {
+	flexibilitySettings, _ := e.PowerConsumptionFlexibilitySettings(entity)
+	currentState, _ := e.State(entity)
+
+	if flexibilitySettings == nil || currentState == nil {
+		return api.ErrDataNotAvailable
+	}
+
+	if !flexibilitySettings.IsStoppable {
```

Comment:
I'm not sure about all those validations.

I think they are good for logging but I can do all the validations under 1 if branch.

WDYT?